File: 2090-number-of-ways-to-arrive-at-destination/number-of-ways-to-arrive-at-destination.cpp
class Solution {
public:
    vector<vector<pair<int, int>>> buildAdj(int n, vector<vector<int>>& roads)
    {
        vector<vector<pair<int, int>>> adj(n);
        for(auto road : roads)
        {
            int u = road[0];
            int v = road[1];
            int time = road[2];
            adj[u].push_back({v, time});
            adj[v].push_back({u, time});
        }
        return adj;
    }

    int countPaths(int n, vector<vector<int>>& roads) 
    {
        int mod = 1e9 + 7;
        auto adj = buildAdj(n, roads);

        vector<int> ways(n, 0);
        vector<long long> dist(n, LLONG_MAX);
        ways[0] = 1;
        dist[0] = 0;

        priority_queue<pair<long long, int>, vector<pair<long long, int>>, greater<>> pq;
        pq.push({0, 0});
        while(!pq.empty())
        {
            auto [currDist, node] = pq.top();
            pq.pop();

            if(currDist > dist[node]) 
            continue;

            for(auto [nbr, wt] : adj[node])
            {
                if(currDist + wt < dist[nbr])
                {
                    dist[nbr] = currDist + wt;
                    ways[nbr] = ways[node];
                    pq.push({dist[nbr], nbr});
                }
                else if(currDist + wt == dist[nbr])
                {
                    ways[nbr] = (ways[nbr] % mod + ways[node] % mod) % mod;
                }
            }
        }
        return ways[n - 1];
    }
};

File: 2090-number-of-ways-to-arrive-at-destination/number-of-ways-to-arrive-at-destination.java
class Solution {

    private final int MOD = 1_000_000_007;
    private List<int[]>[] buildGraph(int n, int[][] roads){
        List<int[]>[] adj = new ArrayList[n];
        for (int i = 0; i < n; i++) {
            adj[i] = new ArrayList<>();
        }
        for(int[] road : roads){
            int node1 = road[0];
            int node2 = road[1];
            int time = road[2];
            adj[node1].add(new int[]{time, node2});
            adj[node2].add(new int[]{time, node1});
        }
        return adj;
    }

    public int countPaths(int n, int[][] roads) {
        var adj = buildGraph(n, roads);
        long[] count = new long[n];
        count[0] = 1;

        long[] dist = new long[n];
        Arrays.fill(dist, Long.MAX_VALUE);
        dist[0] = 0;

        boolean[] visited = new boolean[n];
        // Use Long.compare for safe comparison of long values
        PriorityQueue<long[]> pq = new PriorityQueue<>((a, b) -> Long.compare(a[0], b[0]));
        pq.offer(new long[]{0, 0});
        pq.offer(new long[]{0, 0});
        while (!pq.isEmpty()) {
            long[] curr = pq.poll();
            long currDistance = curr[0];
            int currNode = (int)curr[1];

            if(visited[currNode]) continue;
            visited[currNode] = true;

            for(var adjNode : adj[currNode]){
                int nbr = adjNode[1];
                int weight = adjNode[0];
                if(currDistance + weight < dist[nbr]){
                    dist[nbr] = currDistance + weight;
                    count[nbr] = count[currNode];
                    pq.offer(new long[]{dist[nbr], nbr});
                }
                else if(currDistance + weight == dist[nbr]){
                    count[nbr] = (count[nbr] + count[currNode]) % MOD;
                }
            }
        }
        return (int)count[n - 1] % MOD;
    }
}